File: onyx-database-tests/test/embedded/relationship/RelationshipSelectTest.java
package embedded.relationship;

import category.EmbeddedDatabaseTests;
import com.onyx.exception.OnyxException;
import com.onyx.persistence.query.*;
import embedded.base.BaseTest;
import entities.AddressNoPartition;
import entities.PersonNoPartition;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.experimental.categories.Category;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * Created by tosborn1 on 3/17/17.
 */
@Category({EmbeddedDatabaseTests.class})
public class RelationshipSelectTest extends BaseTest {

    @Before
    public void before() throws OnyxException {
        initialize();
    }

    @After
    public void after() throws IOException {
        shutdown();
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testInvalidQueryException() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }

        Query query = new Query(PersonNoPartition.class, new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart"));
        List<PersonNoPartition> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testDistinctValues() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }

        Query query = new Query(PersonNoPartition.class, new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart"));
        query.setDistinct(true);
        query.setSelections(Arrays.asList("firstName"));
        List<Map> addresses = manager.executeQuery(query);
        assert addresses.size() == 2;
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testQuerySpecificPartition() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }


        QueryCriteria first = new QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian");
        QueryCriteria second = new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(PersonNoPartition.class, first.and(second));
        List<PersonNoPartition> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testSelectAttribute() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }


        QueryCriteria first = new QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian");
        QueryCriteria second = new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(PersonNoPartition.class, first.and(second));
        query.setSelections(Arrays.asList("firstName", "address.street"));
        List<PersonNoPartition> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testSelectRelationship() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }


        QueryCriteria first = new QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian");
        QueryCriteria second = new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(PersonNoPartition.class, first.and(second));
        query.setSelections(Arrays.asList("firstName", "address"));
        List<Map> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
        assert addresses.get(0).get("address") instanceof Map;
        assert ((Map) addresses.get(0).get("address")).get("street").equals("Sluisvaart");

    }

    @Test
    @SuppressWarnings("unchecked")
    public void testToManySelectRelationship() throws OnyxException {
        manager.executeDelete(new Query(PersonNoPartition.class));
        manager.executeDelete(new Query(AddressNoPartition.class));
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);

            PersonNoPartition person2 = new PersonNoPartition();
            person2.firstName = "Timbob";
            person2.lastName = "Rooski" + i;
            person2.address = new AddressNoPartition();
            person2.address.id = person.address.id;
            person2.address.street = "Sluisvaart";
            person2.address.houseNr = 98;
            manager.saveEntity(person2);

            find(person2.address);
        }


        QueryCriteria first = new QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        QueryCriteria second = new QueryCriteria("occupants.firstName", QueryCriteriaOperator.NOT_EQUAL, "Ti!mbob");
        Query query = new Query(AddressNoPartition.class, first.and(second));
        query.setSelections(Arrays.asList("id", "street", "occupants"));

        List<Map> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
        assert addresses.get(0).get("occupants") instanceof List;
        assert ((List) addresses.get(0).get("occupants")).get(0) instanceof Map;
        assert ((Map) ((List) addresses.get(0).get("occupants")).get(0)).get("firstName") != null;
        assert ((Map) ((List) addresses.get(0).get("occupants")).get(1)).get("firstName") != null;
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testToManySelectRelationshipNoRelationshipCriteria() throws OnyxException {
        manager.executeDelete(new Query(PersonNoPartition.class));
        manager.executeDelete(new Query(AddressNoPartition.class));
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);

            PersonNoPartition person2 = new PersonNoPartition();
            person2.firstName = "Timbob";
            person2.lastName = "Rooski" + i;
            person2.address = new AddressNoPartition();
            person2.address.id = person.address.id;
            person2.address.street = "Sluisvaart";
            person2.address.houseNr = 98;
            manager.saveEntity(person2);

            find(person2.address);
        }


        QueryCriteria first = new QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(AddressNoPartition.class, first);
        query.setSelections(Arrays.asList("id", "street", "occupants"));

        List<Map> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
        assert addresses.get(0).get("occupants") instanceof List;
        assert ((List) addresses.get(0).get("occupants")).get(0) instanceof Map;
        assert ((Map) ((List) addresses.get(0).get("occupants")).get(0)).get("firstName") != null;
        assert ((Map) ((List) addresses.get(0).get("occupants")).get(1)).get("firstName") != null;
    }


    @Test
    @SuppressWarnings("unchecked")
    public void testQuerySpecificPartitionOrderBy() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }


        QueryCriteria first = new QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian");
        QueryCriteria second = new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(PersonNoPartition.class, first.and(second));
        query.setQueryOrders(Arrays.asList(new QueryOrder("firstName"), new QueryOrder("address.street")));
        List<PersonNoPartition> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
    }


    @Test
    @SuppressWarnings("unchecked")
    public void testSelectAttributeOrderBy() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }


        QueryCriteria first = new QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian");
        QueryCriteria second = new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(PersonNoPartition.class, first.and(second));
        query.setSelections(Arrays.asList("firstName", "address.street"));
        query.setQueryOrders(Arrays.asList(new QueryOrder("firstName"), new QueryOrder("address.street")));
        List<PersonNoPartition> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testSelectRelationshipOrderBy() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);
        }


        QueryCriteria first = new QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian");
        QueryCriteria second = new QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(PersonNoPartition.class, first.and(second));
        query.setSelections(Arrays.asList("firstName", "address"));
        query.setQueryOrders(Arrays.asList(new QueryOrder("firstName"), new QueryOrder("address.street")));
        List<Map> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
        assert addresses.get(0).get("address") instanceof Map;
        assert ((Map) addresses.get(0).get("address")).get("street").equals("Sluisvaart");

    }

    @Test
    @SuppressWarnings("unchecked")
    public void testToManySelectRelationshipOrderBy() throws OnyxException {
        manager.executeDelete(new Query(PersonNoPartition.class));
        manager.executeDelete(new Query(AddressNoPartition.class));

        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);

            PersonNoPartition person2 = new PersonNoPartition();
            person2.firstName = "Timbob";
            person2.lastName = "Rooski" + i;
            person2.address = new AddressNoPartition();
            person2.address.id = person.address.id;
            person2.address.street = "Sluisvaart";
            person2.address.houseNr = 98;
            manager.saveEntity(person2);

            find(person2.address);
        }


        QueryCriteria first = new QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        QueryCriteria second = new QueryCriteria("occupants.firstName", QueryCriteriaOperator.NOT_EQUAL, "Ti!mbob");
        Query query = new Query(AddressNoPartition.class, first.and(second));
        query.setSelections(Arrays.asList("id", "street", "occupants"));
        query.setQueryOrders(Arrays.asList(new QueryOrder("occupants.firstName")));

        List<Map> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
        assert addresses.get(0).get("occupants") instanceof List;
        assert ((List) addresses.get(0).get("occupants")).get(0) instanceof Map;
        assert ((Map) ((List) addresses.get(0).get("occupants")).get(1)).get("firstName").equals("Timbob");
        assert ((Map) ((List) addresses.get(0).get("occupants")).get(0)).get("firstName").equals("Cristian");
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testToManySelectRelationshipNoRelationshipCriteriaOrderBy() throws OnyxException {
        for (int i = 0; i < 50; i++) {
            PersonNoPartition person = new PersonNoPartition();
            person.firstName = "Cristian";
            person.lastName = "Vogel" + i;
            person.address = new AddressNoPartition();
            person.address.street = "Sluisvaart";
            person.address.houseNr = 98;
            manager.saveEntity(person);

            PersonNoPartition person2 = new PersonNoPartition();
            person2.firstName = "Timbob";
            person2.lastName = "Rooski" + i;
            person2.address = new AddressNoPartition();
            person2.address.id = person.address.id;
            person2.address.street = "Sluisvaart";
            person2.address.houseNr = 98;
            manager.saveEntity(person2);

            find(person2.address);
        }


        QueryCriteria first = new QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart");
        Query query = new Query(AddressNoPartition.class, first);
        query.setSelections(Arrays.asList("id", "street", "occupants"));
        query.setQueryOrders(Arrays.asList(new QueryOrder("street")));

        List<Map> addresses = manager.executeQuery(query);
        assert addresses.size() > 0;
        assert addresses.get(0).get("occupants") instanceof List;
        assert ((List) addresses.get(0).get("occupants")).get(0) instanceof Map;
    }
}

File: onyx-database-tests/test/embedded/relationship/RelationshipSelectTest.kt
package embedded.relationship

import category.EmbeddedDatabaseTests
import com.onyx.exception.OnyxException
import com.onyx.persistence.IManagedEntity
import com.onyx.persistence.query.*
import embedded.base.BaseTest
import entities.AddressNoPartition
import entities.PersonNoPartition
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.experimental.categories.Category

import java.io.IOException
import java.util.Arrays

/**
 * Created by Tim Osborn on 3/17/17.
 */
@Category(EmbeddedDatabaseTests::class)
class RelationshipSelectTest : BaseTest() {

    @Before
    @Throws(OnyxException::class)
    fun before() {
        initialize()
    }

    @After
    @Throws(IOException::class)
    fun after() {
        shutdown()
    }

    @Test
    @Throws(OnyxException::class)
    fun testInvalidQueryException() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }

        val query = Query(PersonNoPartition::class.java, QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart"))
        val addresses = manager.executeQuery<PersonNoPartition>(query)
        assert(addresses.size > 0)
    }

    @Test
    @Throws(OnyxException::class)
    fun testDistinctValues() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }

        val query = Query(PersonNoPartition::class.java, QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart"))
        query.isDistinct = true
        query.selections = Arrays.asList("firstName")
        val addresses = manager.executeQuery<Map<*, *>>(query)
        assert(addresses.size == 2)
    }

    @Test
    @Throws(OnyxException::class)
    fun testQuerySpecificPartition() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }


        val first = QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian")
        val second = QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(PersonNoPartition::class.java, first.and(second))
        val addresses = manager.executeQuery<PersonNoPartition>(query)
        assert(addresses.size > 0)
    }

    @Test
    @Throws(OnyxException::class)
    fun testSelectAttribute() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }


        val first = QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian")
        val second = QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(PersonNoPartition::class.java, first.and(second))
        query.selections = Arrays.asList("firstName", "address.street")
        val addresses = manager.executeQuery<PersonNoPartition>(query)
        assert(addresses.size > 0)
    }

    @Test
    @Throws(OnyxException::class)
    fun testSelectRelationship() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }


        val first = QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian")
        val second = QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(PersonNoPartition::class.java, first.and(second))
        query.selections = Arrays.asList("firstName", "address")
        val addresses = manager.executeQuery<Map<*, *>>(query)
        assert(addresses.size > 0)
        assert(addresses[0]["address"] is Map<*, *>)
        assert((addresses[0]["address"] as Map<*, *>)["street"] == "Sluisvaart")

    }

    @Test
    @Throws(OnyxException::class)
    fun testToManySelectRelationship() {
        manager.executeDelete(Query(PersonNoPartition::class.java))
        manager.executeDelete(Query(AddressNoPartition::class.java))
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)

            val person2 = PersonNoPartition()
            person2.firstName = "Timbob"
            person2.lastName = "Rooski" + i
            person2.address = AddressNoPartition()
            person2.address!!.id = person.address!!.id
            person2.address!!.street = "Sluisvaart"
            person2.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person2)

            find(person2.address!!)
        }


        val first = QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val second = QueryCriteria("occupants.firstName", QueryCriteriaOperator.NOT_EQUAL, "Ti!mbob")
        val query = Query(AddressNoPartition::class.java, first.and(second))
        query.selections = Arrays.asList("id", "street", "occupants")

        val addresses = manager.executeQuery<Map<*, *>>(query)
        assert(addresses.size > 0)
        assert(addresses[0]["occupants"] is List<*>)
        assert((addresses[0]["occupants"] as List<*>)[0] is Map<*, *>)
        assert(((addresses[0]["occupants"] as List<*>)[0] as Map<*, *>)["firstName"] != null)
        assert(((addresses[0]["occupants"] as List<*>)[1] as Map<*, *>)["firstName"] != null)
    }

    @Test
    @Throws(OnyxException::class)
    fun testToManySelectRelationshipNoRelationshipCriteria() {
        manager.executeDelete(Query(PersonNoPartition::class.java))
        manager.executeDelete(Query(AddressNoPartition::class.java))
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)

            val person2 = PersonNoPartition()
            person2.firstName = "Timbob"
            person2.lastName = "Rooski" + i
            person2.address = AddressNoPartition()
            person2.address!!.id = person.address!!.id
            person2.address!!.street = "Sluisvaart"
            person2.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person2)

            find(person2.address!!)
        }


        val first = QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(AddressNoPartition::class.java, first)
        query.selections = Arrays.asList("id", "street", "occupants")

        val addresses = manager.executeQuery<Map<*, *>>(query)
        assert(addresses.size > 0)
        assert(addresses[0]["occupants"] is List<*>)
        assert((addresses[0]["occupants"] as List<*>)[0] is Map<*, *>)
        assert(((addresses[0]["occupants"] as List<*>)[0] as Map<*, *>)["firstName"] != null)
        assert(((addresses[0]["occupants"] as List<*>)[1] as Map<*, *>)["firstName"] != null)
    }


    @Test
    @Throws(OnyxException::class)
    fun testQuerySpecificPartitionOrderBy() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }


        val first = QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian")
        val second = QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(PersonNoPartition::class.java, first.and(second))
        query.queryOrders = Arrays.asList(QueryOrder("firstName"), QueryOrder("address.street"))
        val addresses = manager.executeQuery<PersonNoPartition>(query)
        assert(addresses.size > 0)
    }


    @Test
    @Throws(OnyxException::class)
    fun testSelectAttributeOrderBy() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }


        val first = QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian")
        val second = QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(PersonNoPartition::class.java, first.and(second))
        query.selections = Arrays.asList("firstName", "address.street")
        query.queryOrders = Arrays.asList(QueryOrder("firstName"), QueryOrder("address.street"))
        val addresses = manager.executeQuery<PersonNoPartition>(query)
        assert(addresses.size > 0)
    }

    @Test
    @Throws(OnyxException::class)
    fun testSelectRelationshipOrderBy() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)
        }


        val first = QueryCriteria("firstName", QueryCriteriaOperator.EQUAL, "Cristian")
        val second = QueryCriteria("address.street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(PersonNoPartition::class.java, first.and(second))
        query.selections = Arrays.asList("firstName", "address")
        query.queryOrders = Arrays.asList(QueryOrder("firstName"), QueryOrder("address.street"))
        val addresses = manager.executeQuery<Map<*, *>>(query)
        assert(addresses.size > 0)
        assert(addresses[0]["address"] is Map<*, *>)
        assert((addresses[0]["address"] as Map<*, *>)["street"] == "Sluisvaart")

    }

    @Test
    @Throws(OnyxException::class)
    fun testToManySelectRelationshipOrderBy() {
        manager.executeDelete(Query(PersonNoPartition::class.java))
        manager.executeDelete(Query(AddressNoPartition::class.java))

        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)

            val person2 = PersonNoPartition()
            person2.firstName = "Timbob"
            person2.lastName = "Rooski" + i
            person2.address = AddressNoPartition()
            person2.address!!.id = person.address!!.id
            person2.address!!.street = "Sluisvaart"
            person2.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person2)

            find(person2.address!!)
        }


        val first = QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val second = QueryCriteria("occupants.firstName", QueryCriteriaOperator.NOT_EQUAL, "Ti!mbob")
        val query = Query(AddressNoPartition::class.java, first.and(second))
        query.selections = Arrays.asList("id", "street", "occupants")
        query.queryOrders = Arrays.asList(QueryOrder("occupants.firstName"))

        val addresses = manager.executeQuery<Map<*, *>>(query)
        assert(addresses.size > 0)
        assert(addresses[0]["occupants"] is List<*>)
        assert((addresses[0]["occupants"] as List<*>)[0] is Map<*, *>)
        assert(((addresses[0]["occupants"] as List<*>)[1] as Map<*, *>)["firstName"] == "Timbob")
        assert(((addresses[0]["occupants"] as List<*>)[0] as Map<*, *>)["firstName"] == "Cristian")
    }

    @Test
    @Throws(OnyxException::class)
    fun testToManySelectRelationshipNoRelationshipCriteriaOrderBy() {
        for (i in 0..49) {
            val person = PersonNoPartition()
            person.firstName = "Cristian"
            person.lastName = "Vogel" + i
            person.address = AddressNoPartition()
            person.address!!.street = "Sluisvaart"
            person.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person)

            val person2 = PersonNoPartition()
            person2.firstName = "Timbob"
            person2.lastName = "Rooski" + i
            person2.address = AddressNoPartition()
            person2.address!!.id = person.address!!.id
            person2.address!!.street = "Sluisvaart"
            person2.address!!.houseNr = 98
            manager.saveEntity<IManagedEntity>(person2)

            find(person2.address!!)
        }


        val first = QueryCriteria("street", QueryCriteriaOperator.EQUAL, "Sluisvaart")
        val query = Query(AddressNoPartition::class.java, first)
        query.selections = Arrays.asList("id", "street", "occupants")
        query.queryOrders = Arrays.asList(QueryOrder("street"))

        val addresses = manager.executeQuery<Map<*, *>>(query)
        assert(addresses.size > 0)
        assert(addresses[0]["occupants"] is List<*>)
        assert((addresses[0]["occupants"] as List<*>)[0] is Map<*, *>)
    }
}
